apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerClientFactory.cs

```diff
@@ -0,0 +1,97 @@
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Threading.Tasks;
+using Google.Api.Gax;
+using Google.Api.Gax.Grpc;
+using Google.Apis.Auth.OAuth2;
+using Google.Cloud.Spanner.V1;
+using Google.Cloud.Spanner.V1.Logging;
+using Grpc.Auth;
+using Grpc.Core;
+
+namespace Google.Cloud.Spanner.Data
+{
+    internal interface ISpannerClientFactory
+    {
+        Task<SpannerClient> CreateClientAsync(ServiceEndpoint endpoint, ITokenAccess credential);
+    }
+
+    /// <summary>
+    /// For mockability and testability of the <see cref="ClientPool"/> class.
+    /// </summary>
+    internal class SpannerClientFactory : ISpannerClientFactory
+    {
+        internal static readonly ISpannerClientFactory Default = new SpannerClientFactory();
+
+        private SpannerClientFactory()
+        {
+        }
+
+        private static async Task<ChannelCredentials> CreateDefaultChannelCredentialsAsync()
+        {
+            var appDefaultCredentials = await GoogleCredential.GetApplicationDefaultAsync().ConfigureAwait(false);
+            if (appDefaultCredentials.IsCreateScopedRequired)
+            {
+                appDefaultCredentials = appDefaultCredentials.CreateScoped(SpannerClient.DefaultScopes);
+            }
+            return appDefaultCredentials.ToChannelCredentials();
+        }
+
+        /// <inheritdoc />
+        public async Task<SpannerClient> CreateClientAsync(ServiceEndpoint endpoint, ITokenAccess credential)
+        {
+            ChannelCredentials channelCredentials;
+
+            if (credential == null)
+            {
+                channelCredentials = await CreateDefaultChannelCredentialsAsync().ConfigureAwait(false);
+            }
+            else
+            {
+                channelCredentials = credential.ToChannelCredentials();
+            }
+
+            var channel = new Channel(
+                endpoint.Host,
+                endpoint.Port,
+                channelCredentials);
+            Logger.LogPerformanceCounterFn("SpannerClient.RawCreateCount", x => x + 1);
+
+            //Pull the timeout from spanner options.
+            //The option must be set before OpenAsync is called.
+            var callSettings = CallSettings.FromCallTiming(
+                CallTiming.FromRetry(
+                    new RetrySettings(
+                        SpannerSettings.GetDefaultRetryBackoff(),
+                        SpannerSettings.GetDefaultTimeoutBackoff(),
+                        Expiration.FromTimeout(SpannerOptions.Instance.Timeout),
+                        SpannerSettings.NonIdempotentRetryFilter
+                    )));
+
+            return SpannerClient.Create(
```

Comment:
I think we chatted on this briefly.  definitely seems harder than it should be to set an overall timeout.  I should not have to know to specifics about how to construct the entire retry settings.

It would be much better if the client call settings werent completely generated code in vkit but was in some way part of a base framework with overrideable/mutable properties.  I know you are big on immutability here as a way of ensuring that developers better understand when settings are frozen, but in the case of retry settings, you don't actually need to freeze them until you are inside a call.
So what I would do is make all of the retrysettings mutable on (BASE)Client and freeze them into immutable properties only when a callcontext is constructed (an rpc call is underway).